add replica set option

StefanoDeVuono opened the pull request:
This PR is to allow users to add replica sets to Mongo in GitHub Actions.

We use the `INPUT` environment variable to name a replica set if there is one and then start docker with a replica set and initialize it.

TODO:
- [x] test that it works as an actual GitHub workflow (I've simulated it with `nektos/act` locally, but that meant editing the files a bit, so it may not be an accurate test)
- ~~[ ] we might want to refactor variables. Right now, we mix named variables and numbered arguments, but it might be easier to move to named variables for everything.~~

StefanoDeVuono commented on start-mongodb.sh:
```diff
@@ -1,3 +1,15 @@
+fi
+
+docker run --name mongodb --publish 27017:27017 --detach mongo:$1 $REPL_COMMAND
+
```
I kinda don't like the mixing of named shell variables and numbered args. Maybe we can move towards named variables in general?

marcuspoehls commented:
@StefanoDeVuono Hey Stefano, thank you for the PR! I'll look over the changes and reply back. I'll probably be able to put my thoughts into this by tomorrow.

marcuspoehls commented:
Hey Stefano. Thank you for your patience!

I'm currently busy with work. I'm sure I'll have time to get back to this PR and your work in the next days. Probably on Wednesday.

StefanoDeVuono commented:
@marcuspoehls I'm still testing it out. I'll squash and let you know what I find

StefanoDeVuono commented:
@marcuspoehls The code should be good now.

I created an automated test. It's in node.js, but lemme know if you want me to add it.

marcuspoehls commented on start-mongodb.sh:
```diff
@@ -1,3 +1,18 @@
+
+if [ -n "$MONGO_REPL_SET" ]
+then
+    sleep 1
```
Can you please explain why you added this `sleep 1` here?

StefanoDeVuono replied:
I think it's from an old recipe to make sure the server is started before initiating the replica set.

Let's take it out and see if it still works?

marcuspoehls commented on action.yml:
```diff
@@ -1,18 +1,18 @@
     required: false
-    default: 'latest'
+    default: "latest"
 
```
Is the REPL_SET input missing here?

StefanoDeVuono replied:
We use `env`. I'll update the Readme with an example.

marcuspoehls commented:
@StefanoDeVuono Hey Stefano, thank you for your comments on my questions. I can dedicate some time today to get this replica set feature added to this GitHub Action.

Can you please add your Node.js test to this PR? I think it’s a good idea to have an extra test for the replica set.

In the meantime, I’ll read about MongoDB replica sets and how we can ensure the functionality is working as expected. You did a great job extending the existing code with only a few lines to support replica sets. The best start I can get :) Thank you!

StefanoDeVuono commented:
I added tests!

And I added the sleep back in. it looks like the server won't be ready for the `rs.initiate()` command otherwise. See here: https://github.com/StefanoDeVuono/mongodb-github-action/pull/2/checks?check_run_id=538467496

And with sleep:
https://github.com/StefanoDeVuono/mongodb-github-action/pull/2/checks?sha=64910570108c1cf469c729acebf1037fbd37d68d

marcuspoehls commented:
@StefanoDeVuono Thank you for the test.

I think your replica set isn't testing against the replica set. You're not passing down the `options` containing the replica set name when connecting Mongoose to MongoDB.

I was toying with replica set support in the [`replica-sets`](https://github.com/supercharge/mongodb-github-action/tree/replica-sets) branch. I wasn't able to make a connection to the MongoDB replica set in the docker container.

Maybe you have an idea. Once we get that fixed, I'll merge this PR.

Thank you for your help!

StefanoDeVuono commented:
@marcuspoehls I got it!

The replica sets have a `members.name` field which is the hostname. Output of `rs.status()` is 	

```
"set": "rsTest",
"...": "...",
"members" : [
    {
        "_id" : 0,
        "name" : "a7f90e76379b:27017",
        "...": "...",
```
on a dockerized replica set initialized with `rs.initiate()` where `a7f90e76379b` is the host, `27017` is the port, a `set` is the name of the replica set.

So, I pass in a basic replica set config `rs.initiate({ \"_id\": \"$MONGO_REPL_SET\", \"members\": [{ \"_id\": 0, \"host\": \"localhost\" }] })` and it seems to work

StefanoDeVuono commented on test.js:
```diff
@@ -0,0 +1,61 @@
+  }
+
+test.before('setup', async t => {
+  const [err, _res] = await to(mongoose.connect(mongoUri, options))
```
It's active now!

StefanoDeVuono commented on start-mongodb.sh:
```diff
@@ -1,3 +1,18 @@
+then
+    echo "initiating $MONGO_REPL_SET."
+    sleep 1
+    docker exec -t mongodb mongo --eval "rs.initiate({ \"_id\": \"$MONGO_REPL_SET\", \"members\": [{ \"_id\": 0, \"host\": \"localhost\" }] })"
```
The secret sauce.

I was able to connect to the dockerized when I didn't create a replica set which pointed to the fact that the replica set initialization did some network stuff. Indeed the `rs.initiate()` command seems to bind an entry point to the replica set on a certain hostname which is the container id (`ef2bdf8b4079`)

```
$ mongo --eval "rs.status()"
...
members" : [
                {
                        "_id" : 0,
                        "name" : "ef2bdf8b4079:27017",
...

$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                      NAMES
ef2bdf8b4079        mongo:4.2           "docker-entrypoint.s…"   22 seconds ago      Up 20 seconds       0.0.0.0:27017->27017/tcp   mongodb

marcuspoehls replied:
Yes, good catch! Using `host: 'localhost'` fixes the networking issue! Thank you

marcuspoehls commented:
@StefanoDeVuono Thank you for your work Stefano! I really appreciate your help. I’ll release a new version by the end of today. I’m currently on vacation and will squeeze in some time to push a new release. Again, thank you for your patience and walking through all the issues related to replica sets 🙌
